In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
train = pd.read_csv(r'C:\Users\abida\Documents\GitHub\Credit_Risk--Probability_of_default\cs-training.csv').drop(['Unnamed: 0'],axis=1)
test = pd.read_csv(r'C:\Users\abida\Documents\GitHub\Credit_Risk--Probability_of_default\cs-test.csv').drop(['Unnamed: 0'],axis=1)

# 1- Data Exploration:

In [3]:
train.head()

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
0                 1                              0.766127   45   
1                 0                              0.957151   40   
2                 0                              0.658180   38   
3                 0                              0.233810   30   
4                 0                              0.907239   49   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                     2   0.802982         9120.0   
1                                     0   0.121876         2600.0   
2                                     1   0.085113         3042.0   
3                                     0   0.036050         3300.0   
4                                     1   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                               13                        0   
1                                4                        0   
2                                2                        1   
3                                5                        0   
4                                7                        0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                             6                                     0   
1                             0                                     0   
2                             0                                     0   
3                             0                                     0   
4                             1                                     0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

In [4]:
# .duplicated returns True/False if the entire row is duplicated
train.duplicated()

0         False
1         False
2         False
3         False
4         False
          ...  
149995    False
149996    False
149997    False
149998    False
149999    False
Length: 150000, dtype: bool

In [5]:
#to see how many duplicates we will simply use value_counts().
train.duplicated().value_counts()

False    149391
True        609
Name: count, dtype: int64

we have 609 records that are duplicates. i will just delete them as initially i'm looking for my model to train on unique cases.

In [6]:
train_redup = train.drop_duplicates()

In [7]:
#as you can see we no longer have duplicates
train_redup.duplicated().value_counts()

False    149391
Name: count, dtype: int64

I will first take a look on what does the training data look like

In [8]:
train_redup.head()

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
0                 1                              0.766127   45   
1                 0                              0.957151   40   
2                 0                              0.658180   38   
3                 0                              0.233810   30   
4                 0                              0.907239   49   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                     2   0.802982         9120.0   
1                                     0   0.121876         2600.0   
2                                     1   0.085113         3042.0   
3                                     0   0.036050         3300.0   
4                                     1   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                               13                        0   
1                                4                        0   
2                                2                        1   
3                                5                        0   
4                                7                        0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                             6                                     0   
1                             0                                     0   
2                             0                                     0   
3                             0                                     0   
4                             1                                     0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

let's take a look on the shape of the training data as well as the test data

In [9]:
train_redup.shape


(149391, 11)

In [10]:
test.shape

(101503, 11)

let's see if we have null values.

In [11]:
train_redup.isnull().sum()

SeriousDlqin2yrs                            0
RevolvingUtilizationOfUnsecuredLines        0
age                                         0
NumberOfTime30-59DaysPastDueNotWorse        0
DebtRatio                                   0
MonthlyIncome                           29221
NumberOfOpenCreditLinesAndLoans             0
NumberOfTimes90DaysLate                     0
NumberRealEstateLoansOrLines                0
NumberOfTime60-89DaysPastDueNotWorse        0
NumberOfDependents                       3828
dtype: int64

In [12]:
test.isnull().sum()

SeriousDlqin2yrs                        101503
RevolvingUtilizationOfUnsecuredLines         0
age                                          0
NumberOfTime30-59DaysPastDueNotWorse         0
DebtRatio                                    0
MonthlyIncome                            20103
NumberOfOpenCreditLinesAndLoans              0
NumberOfTimes90DaysLate                      0
NumberRealEstateLoansOrLines                 0
NumberOfTime60-89DaysPastDueNotWorse         0
NumberOfDependents                        2626
dtype: int64

first we will look into the null values of number of Dependents and see what we can do about it.

In [13]:
train_redup[train_redup.NumberOfDependents.isnull()]

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
8                      0                              0.116951   27   
96                     0                              0.542243   48   
109                    0                              0.041258   61   
159                    0                              0.000000   63   
238                    0                              1.000000   28   
...                  ...                                   ...  ...   
149826                 0                              0.027319   71   
149854                 0                              0.056589   75   
149894                 0                              0.017904   55   
149948                 0                              0.055530   27   
149965                 0                              1.000000   27   

        NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
8                                          0       46.0            NaN   
96                                         2       10.0            NaN   
109                                        0     4739.0            NaN   
159                                        0        2.0            NaN   
238                                        0        0.0            NaN   
...                                      ...        ...            ...   
149826                                     0     1419.0            NaN   
149854                                     0        8.0            NaN   
149894                                     0     1058.0            NaN   
149948                                     0       10.0            NaN   
149965                                     0       98.0            NaN   

        NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
8                                     2                        0   
96                                    2                        0   
109                                  11                        0   
159                                   4                        0   
238                                   0                        0   
...                                 ...                      ...   
149826                               13                        0   
149854                                3                        0   
149894                                8                        0   
149948                                1                        0   
149965                                2                        0   

        NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
8                                  0                                     0   
96                                 0                                     0   
109                                4                                     0   
159                                0                                     0   
238                                0                                     0   
...                              ...                                   ...   
149826                             1                                     0   
149854                             0                                     0   
149894                             1                                     1   
149948                             0                                     0   
149965                             0                                     0   

        NumberOfDependents  
8                      NaN  
96                     NaN  
109                    NaN  
159                    NaN  
238                    NaN  
...                    ...  
149826                 NaN  
149854                 NaN  
149894                 NaN  
149948                 NaN  
149965                 NaN  

[3828 rows x 11 columns]

In [14]:
train_redup[train_redup.NumberOfDependents==0]

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
2                      0                              0.658180   38   
3                      0                              0.233810   30   
4                      0                              0.907239   49   
6                      0                              0.305682   57   
7                      0                              0.754464   39   
...                  ...                                   ...  ...   
149994                 0                              0.385742   50   
149995                 0                              0.040674   74   
149997                 0                              0.246044   58   
149998                 0                              0.000000   30   
149999                 0                              0.850283   64   

        NumberOfTime30-59DaysPastDueNotWorse    DebtRatio  MonthlyIncome  \
2                                          1     0.085113         3042.0   
3                                          0     0.036050         3300.0   
4                                          1     0.024926        63588.0   
6                                          0  5710.000000            NaN   
7                                          0     0.209940         3500.0   
...                                      ...          ...            ...   
149994                                     0     0.404293         3400.0   
149995                                     0     0.225131         2100.0   
149997                                     0  3870.000000            NaN   
149998                                     0     0.000000         5716.0   
149999                                     0     0.249908         8158.0   

        NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
2                                     2                        1   
3                                     5                        0   
4                                     7                        0   
6                                     8                        0   
7                                     8                        0   
...                                 ...                      ...   
149994                                7                        0   
149995                                4                        0   
149997                               18                        0   
149998                                4                        0   
149999                                8                        0   

        NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
2                                  0                                     0   
3                                  0                                     0   
4                                  1                                     0   
6                                  3                                     0   
7                                  0                                     0   
...                              ...                                   ...   
149994                             0                                     0   
149995                             1                                     0   
149997                             1                                     0   
149998                             0                                     0   
149999                             2                                     0   

        NumberOfDependents  
2                      0.0  
3                      0.0  
4                      0.0  
6                      0.0  
7                      0.0  
...                    ...  
149994                 0.0  
149995                 0.0  
149997                 0.0  
149998                 0.0  
149999                 0.0  

[86392 rows x 11 columns]

As you can see my first instinct was to see if that just meant that the borrower had no dependents that was wrong in that case we wouldn't have had any records where the number of dependents is equal to 0. so in this case i will try to find the best way to replace the missing values to get the best results.

In [16]:
train_redup[train_redup.NumberOfDependents.isnull()].describe()

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines          age  \
count       3828.000000                           3828.000000  3828.000000   
mean           0.046499                             11.003369    59.741641   
std            0.210592                            240.656436    18.345175   
min            0.000000                              0.000000    21.000000   
25%            0.000000                              0.009228    48.000000   
50%            0.000000                              0.047670    61.000000   
75%            0.000000                              0.259028    74.000000   
max            1.000000                          10821.000000   109.000000   

       NumberOfTime30-59DaysPastDueNotWorse      DebtRatio  MonthlyIncome  \
count                           3828.000000    3828.000000            0.0   
mean                               0.572623    1110.713689            NaN   
std                                6.527320    4235.410634            NaN   
min                                0.000000       0.000000            NaN   
25%                                0.000000      25.000000            NaN   
50%                                0.000000     398.000000            NaN   
75%                                0.000000    1587.000000            NaN   
max                               98.000000  220516.000000            NaN   

       NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
count                      3828.000000              3828.000000   
mean                          5.708986                 0.497126   
std                           4.086337                 6.523199   
min                           0.000000                 0.000000   
25%                           3.000000                 0.000000   
50%                           5.000000                 0.000000   
75%                           8.000000                 0.000000   
max                          30.000000                98.000000   

       NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
count                   3828.000000                           3828.000000   
mean                       0.605799                              0.474138   
std                        0.920990                              6.520343   
min                        0.000000                              0.000000   
25%                        0.000000                              0.000000   
50%                        0.000000                              0.000000   
75%                        1.000000                              0.000000   
max                       15.000000                             98.000000   

       NumberOfDependents  
count                 0.0  
mean                  NaN  
std                   NaN  
min                   NaN  
25%                   NaN  
50%                   NaN  
75%                   NaN  
max                   NaN

In [17]:
train_redup.describe()

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines            age  \
count     149391.000000                         149391.000000  149391.000000   
mean           0.066999                              6.071087      52.306237   
std            0.250021                            250.263672      14.725962   
min            0.000000                              0.000000       0.000000   
25%            0.000000                              0.030132      41.000000   
50%            0.000000                              0.154235      52.000000   
75%            0.000000                              0.556494      63.000000   
max            1.000000                          50708.000000     109.000000   

       NumberOfTime30-59DaysPastDueNotWorse      DebtRatio  MonthlyIncome  \
count                         149391.000000  149391.000000   1.201700e+05   
mean                               0.393886     354.436740   6.675098e+03   
std                                3.852953    2041.843455   1.438958e+04   
min                                0.000000       0.000000   0.000000e+00   
25%                                0.000000       0.177441   3.400000e+03   
50%                                0.000000       0.368234   5.400000e+03   
75%                                0.000000       0.875279   8.250000e+03   
max                               98.000000  329664.000000   3.008750e+06   

       NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
count                    149391.000000            149391.000000   
mean                          8.480892                 0.238120   
std                           5.136515                 3.826165   
min                           0.000000                 0.000000   
25%                           5.000000                 0.000000   
50%                           8.000000                 0.000000   
75%                          11.000000                 0.000000   
max                          58.000000                98.000000   

       NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
count                 149391.000000                         149391.000000   
mean                       1.022391                              0.212503   
std                        1.130196                              3.810523   
min                        0.000000                              0.000000   
25%                        0.000000                              0.000000   
50%                        1.000000                              0.000000   
75%                        2.000000                              0.000000   
max                       54.000000                             98.000000   

       NumberOfDependents  
count       145563.000000  
mean             0.759863  
std              1.116141  
min              0.000000  
25%              0.000000  
50%              0.000000  
75%              1.000000  
max             20.000000

as you can see records with missing values of Number of dependents have less people who experienced 90 days past due delinquency or worse.

In [20]:
train_redup["NumberOfDependents"].agg(['mode'])

mode
0   0.0

In [21]:
train_redup.groupby(['NumberOfDependents']).size()

NumberOfDependents
0.0     86392
1.0     26314
2.0     19521
3.0      9483
4.0      2862
5.0       746
6.0       158
7.0        51
8.0        24
9.0         5
10.0        5
13.0        1
20.0        1
dtype: int64

most of the past customers do not have any dependents.

so we will fill the missing values of this variable(NumberOfDependents) with 0 and see what's that gonna look like.

In [26]:
fam_miss = train_redup[train_redup.NumberOfDependents.isnull()]
fam_nmiss = train_redup[train_redup.NumberOfDependents.notnull()]

In [27]:
fam_miss['NumberOfDependents'] = fam_miss['NumberOfDependents'].fillna(0)

C:\Users\abida\AppData\Local\Temp\ipykernel_11364\3489319309.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  fam_miss['NumberOfDependents'] = fam_miss['NumberOfDependents'].fillna(0)


In [30]:
fam_miss.isnull().sum()

SeriousDlqin2yrs                           0
RevolvingUtilizationOfUnsecuredLines       0
age                                        0
NumberOfTime30-59DaysPastDueNotWorse       0
DebtRatio                                  0
MonthlyIncome                           3828
NumberOfOpenCreditLinesAndLoans            0
NumberOfTimes90DaysLate                    0
NumberRealEstateLoansOrLines               0
NumberOfTime60-89DaysPastDueNotWorse       0
NumberOfDependents                         0
dtype: int64

----------------------------

Same with the monthly income.


In [23]:
train_redup[train_redup.MonthlyIncome.isnull()]

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
6                      0                              0.305682   57   
8                      0                              0.116951   27   
16                     0                              0.061086   78   
32                     0                              0.083418   62   
41                     0                              0.072898   81   
...                  ...                                   ...  ...   
149976                 0                              0.000627   76   
149977                 0                              0.236450   29   
149984                 0                              0.037548   84   
149992                 0                              0.871976   50   
149997                 0                              0.246044   58   

        NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
6                                          0     5710.0            NaN   
8                                          0       46.0            NaN   
16                                         0     2058.0            NaN   
32                                         0      977.0            NaN   
41                                         0       75.0            NaN   
...                                      ...        ...            ...   
149976                                     0       60.0            NaN   
149977                                     0      349.0            NaN   
149984                                     0       25.0            NaN   
149992                                     0     4132.0            NaN   
149997                                     0     3870.0            NaN   

        NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
6                                     8                        0   
8                                     2                        0   
16                                   10                        0   
32                                    6                        0   
41                                    7                        0   
...                                 ...                      ...   
149976                                5                        0   
149977                                3                        0   
149984                                5                        0   
149992                               11                        0   
149997                               18                        0   

        NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
6                                  3                                     0   
8                                  0                                     0   
16                                 2                                     0   
32                                 1                                     0   
41                                 0                                     0   
...                              ...                                   ...   
149976                             0                                     0   
149977                             0                                     0   
149984                             0                                     0   
149992                             1                                     0   
149997                             1                                     0   

        NumberOfDependents  
6                      0.0  
8                      NaN  
16                     0.0  
32                     0.0  
41                     0.0  
...                    ...  
149976                 0.0  
149977                 0.0  
149984                 0.0  
149992                 3.0  
149997                 0.0  

[29221 rows x 11 columns]

In [24]:
train_redup[train_redup.MonthlyIncome==0]

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
14                     0                              0.019657   76   
50                     0                              0.818978   73   
73                     0                              0.059669   31   
90                     0                              0.039388   51   
298                    0                              0.085152   25   
...                  ...                                   ...  ...   
149383                 0                              0.239302   63   
149562                 0                              0.230799   72   
149646                 0                              0.080084   69   
149673                 0                              0.132362   82   
149950                 0                              0.004827   61   

        NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
14                                         0      477.0            0.0   
50                                         0     3095.0            0.0   
73                                         0     3162.0            0.0   
90                                         0    15466.0            0.0   
298                                        0     1005.0            0.0   
...                                      ...        ...            ...   
149383                                     0     3153.0            0.0   
149562                                     0      430.0            0.0   
149646                                     0     1248.0            0.0   
149673                                     0     4123.0            0.0   
149950                                     0     1940.0            0.0   

        NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
14                                    6                        0   
50                                    9                        0   
73                                   11                        0   
90                                    7                        0   
298                                   5                        0   
...                                 ...                      ...   
149383                               29                        0   
149562                                7                        0   
149646                                8                        0   
149673                               14                        0   
149950                               10                        0   

        NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
14                                 1                                     0   
50                                 1                                     1   
73                                 2                                     0   
90                                 0                                     0   
298                                0                                     0   
...                              ...                                   ...   
149383                             1                                     0   
149562                             0                                     0   
149646                             1                                     0   
149673                             2                                     0   
149950                             1                                     0   

        NumberOfDependents  
14                     0.0  
50                     0.0  
73                     1.0  
90                     0.0  
298                    0.0  
...                    ...  
149383                 0.0  
149562                 0.0  
149646                 1.0  
149673                 0.0  
149950                 0.0  

[1616 rows x 11 columns]

In [25]:
train_redup[train_redup.MonthlyIncome==0].describe()

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines          age  \
count       1616.000000                           1616.000000  1616.000000   
mean           0.040223                              4.336369    48.620668   
std            0.196542                            147.433057    16.558826   
min            0.000000                              0.000000    21.000000   
25%            0.000000                              0.020273    36.000000   
50%            0.000000                              0.107736    47.000000   
75%            0.000000                              0.472699    62.000000   
max            1.000000                           5893.000000    97.000000   

       NumberOfTime30-59DaysPastDueNotWorse     DebtRatio  MonthlyIncome  \
count                           1616.000000   1616.000000         1616.0   
mean                               0.421411   1591.094678            0.0   
std                                4.896579   2828.748394            0.0   
min                                0.000000      0.000000            0.0   
25%                                0.000000    110.500000            0.0   
50%                                0.000000    946.000000            0.0   
75%                                0.000000   2204.000000            0.0   
max                               98.000000  60212.000000            0.0   

       NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
count                      1616.000000              1616.000000   
mean                          7.131188                 0.309406   
std                           4.831283                 4.890459   
min                           0.000000                 0.000000   
25%                           3.000000                 0.000000   
50%                           6.000000                 0.000000   
75%                          10.000000                 0.000000   
max                          31.000000                98.000000   

       NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
count                   1616.000000                           1616.000000   
mean                       0.724629                              0.287129   
std                        0.905593                              4.881175   
min                        0.000000                              0.000000   
25%                        0.000000                              0.000000   
50%                        0.000000                              0.000000   
75%                        1.000000                              0.000000   
max                        9.000000                             98.000000   

       NumberOfDependents  
count         1616.000000  
mean             0.738861  
std              1.190728  
min              0.000000  
25%              0.000000  
50%              0.000000  
75%              1.000000  
max             10.000000

In [32]:
fam_miss[fam_miss.MonthlyIncome.isnull()].describe()

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines          age  \
count       3828.000000                           3828.000000  3828.000000   
mean           0.046499                             11.003369    59.741641   
std            0.210592                            240.656436    18.345175   
min            0.000000                              0.000000    21.000000   
25%            0.000000                              0.009228    48.000000   
50%            0.000000                              0.047670    61.000000   
75%            0.000000                              0.259028    74.000000   
max            1.000000                          10821.000000   109.000000   

       NumberOfTime30-59DaysPastDueNotWorse      DebtRatio  MonthlyIncome  \
count                           3828.000000    3828.000000            0.0   
mean                               0.572623    1110.713689            NaN   
std                                6.527320    4235.410634            NaN   
min                                0.000000       0.000000            NaN   
25%                                0.000000      25.000000            NaN   
50%                                0.000000     398.000000            NaN   
75%                                0.000000    1587.000000            NaN   
max                               98.000000  220516.000000            NaN   

       NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
count                      3828.000000              3828.000000   
mean                          5.708986                 0.497126   
std                           4.086337                 6.523199   
min                           0.000000                 0.000000   
25%                           3.000000                 0.000000   
50%                           5.000000                 0.000000   
75%                           8.000000                 0.000000   
max                          30.000000                98.000000   

       NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
count                   3828.000000                           3828.000000   
mean                       0.605799                              0.474138   
std                        0.920990                              6.520343   
min                        0.000000                              0.000000   
25%                        0.000000                              0.000000   
50%                        0.000000                              0.000000   
75%                        1.000000                              0.000000   
max                       15.000000                             98.000000   

       NumberOfDependents  
count              3828.0  
mean                  0.0  
std                   0.0  
min                   0.0  
25%                   0.0  
50%                   0.0  
75%                   0.0  
max                   0.0

In [34]:
fam_miss['MonthlyIncome'].agg(['mean','median','min','max'])

c:\Users\abida\AppData\Local\Programs\Python\Python312\Lib\site-packages\numpy\lib\nanfunctions.py:1215: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)


mean     NaN
median   NaN
min      NaN
max      NaN
Name: MonthlyIncome, dtype: float64

### _Hypothesis 1:_
Let's assume the worst case possible for people who have missing values of monthly income is because they want to hide the fact that it's 0 (for a better deal maybe? idk)